### PR TITLE
Support FP8 Quantization and Inference Run on Intel Gaudi (HPU) using INC (Intel Neural Compressor)

### PR DESCRIPTION
**Pull request by nirda7**
This PR adds support to FP8 Quantization and Inference run on Intel Gaudi (HPU) using INC ([Intel Neural Compressor](https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Quantization/index.html#quantization-prerequisites)).
Currently, quantization is validated only in Llama models.
 
* Measurement files are required to run quantized models with vLLM on Gaudi accelerators. The FP8 model calibration procedure is described in the vllm-hpu-extention package: https://github.com/HabanaAI/vllm-hpu-extension/tree/main/calibration/README.md
Measurements are device dependent - Don't use measurements collected on Gaudi3 on Gaudi2 accelerators as it might cause accuracy issues.

Running Inference in FP8 with INC:
Specify quantization method "inc" and kv cache dtype "fp8_inc" as parameters to the the LLM object.
It will require to set an environment variable "QUANT_CONFIG" which will point to a 'JSON config file' (https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Quantization/Inference_Using_FP8.html#supported-json-config-file-options) in QUANTIZE mode. Make sure there are measurement files/scale files in the folder specified as the "dump_stats_path" in the json config file. (If none exists, scale files are generated during the inference run using the measurement files)
At the end of the run, the model executor's shutdown method must be called.
 
More information on vLLM quantization using INC will be shown here (added in this PR): https://github.com/vllm-project/vllm/blob/main/docs/source/features/quantization/inc.md
 
This PR adds a new flag "weights_load_device" which allows uploading the model's (unquantized) weights onto a different device than the device on which the model will run. If not provided the behavior is kept by using the device specified in the device config.

### CODE REVIEW

File: vllm/worker/hpu_model_runner.py [hunk 1]
```diff
@@ -274,17 +274,24 @@ def precompute_indices_and_offsets(block_size, slot_mapping, is_prompt):
     return indices, offsets
 
 
-def modify_decoder_layer(module: torch.nn.Module, suffix="DecoderLayer"):
+def modify_decoder_layer(module: torch.nn.Module,
```
[the remark this answers is not in the document]
Comment (nirda7) [reply vllm/worker/hpu_model_runner.py:277]: @zhouyu5 
It is used for both BF16 and FP8.
I just added another mark-step before the first layer (as we used to run in vllm-fork)

File: vllm/executor/ray_distributed_executor.py [hunk 2]
```diff
@@ -100,11 +100,21 @@ def _init_executor(self) -> None:
...
             "Shutting down Ray distributed executor. If you see error log "
             "from logging.cc regarding SIGTERM received, please ignore because "
             "this is the expected termination process in Ray.")
+        if getattr(self, 'shutdown_workers', False):
+            self._run_workers("shutdown")
+            self.shutdown_workers = False
+        if getattr(self, 'terminate_ray', False):
+            for worker in self.workers:
+                worker.__ray_terminate__.remote()
+            self.terminate_ray = False
```
Comment (mgoin) [inline vllm/executor/ray_distributed_executor.py:117]: Why do you need to add shutdown routines to all the executors? This should be taken care of already. cc @comaniac for the changes to Ray

File: vllm/worker/hpu_worker.py [hunk 3]
```diff
@@ -469,12 +466,13 @@ def _allocate_kv_cache(
         kv_cache_shape = self.attn_backend.get_kv_cache_shape(
             num_blocks, self.block_size, self.num_kv_heads, self.head_size)
         kv_cache: List[Tuple[torch.Tensor, torch.Tensor]] = []
+        dtype = self.dtype
+        if device != 'hpu' and self.dtype == torch.float8_e4m3fn:
+            dtype = torch.uint8
```
Comment (mgoin) [inline vllm/worker/hpu_worker.py:471]: What is the purpose of seeing a new kv_cache_dtype `"fp8_inc": torch.float8_e4m3fn` just to end up using the same torch.uint8 dtype?

File: vllm/model_executor/layers/quantization/inc.py
Comment (mgoin) [inline vllm/model_executor/layers/quantization/inc.py:1]: What is the purpose of adding this quantization class if it seems to do nothing?

File: vllm/engine/arg_utils.py [hunk 4]
```diff
@@ -336,6 +337,12 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
             'Model Streamer \n'
             '* "bitsandbytes" will load the weights using bitsandbytes '
             'quantization.\n')
+        parser.add_argument("--weights-load-device",
+                            type=str,
+                            default=EngineArgs.weights_load_device,
+                            choices=DEVICE_OPTIONS,
+                            help='Device to which model weights '
+                            'will be loaded.')
```
Comment (mgoin) [inline vllm/engine/arg_utils.py:345]: Why do you need to load the weights on CPU? How do they end up on the device? This seems like a feature that will be difficult to support for the many different backends

File: vllm/model_executor/layers/quantization/inc.py
Comment (nirda7) [reply vllm/model_executor/layers/quantization/inc.py:1]: We want to be compatible with the API - we have some definitions there complying with INC quantization method.

File: vllm/executor/ray_distributed_executor.py (hunk 2 above)
Comment (nirda7) [reply vllm/executor/ray_distributed_executor.py]: This would be only part of Inc flow:
For Inc quantization method we need this shutdown process to prevent the processing from hanging up and to assure complete ending for the use of Inc.
Based on Inc implementation this would be the best way to shutdown the process.

**Comment by simon-mo** [comment]
@mgoin can you help make a pass here as code owners for quantization methods?

File: vllm/executor/ray_distributed_executor.py [hunk 5]
```diff
@@ -123,11 +123,22 @@ def _init_executor(self) -> None:
...
 
+        self.shutdown_workers = True
+        self.terminate_ray = True
+
     def shutdown(self) -> None:
         logger.info(
             "Shutting down Ray distributed executor. If you see error log "
             "from logging.cc regarding SIGTERM received, please ignore because "
             "this is the expected termination process in Ray.")
+        if (self.model_config.quantization == 'inc'):
```
Comment (mgoin) [inline vllm/executor/ray_distributed_executor.py:134]: Why do you need different distributed behavior when inc quantization is enabled?

File: vllm/model_executor/layers/quantization/inc.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    @classmethod
+    def from_config(cls, config: Dict[str, Any]) -> "INCConfig":
+        raise AssertionError
+
+    def get_quant_method(self, layer: torch.nn.Module,
+                         prefix: str) -> Optional["QuantizeMethodBase"]:
+        if isinstance(layer, LinearBase):
+            return UnquantizedLinearMethod()
+        elif isinstance(layer, FusedMoE):
+            return UnquantizedFusedMoEMethod()
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/inc.py:36]: Why is this quantization layer just producing unquantized methods?

File: vllm/model_executor/layers/linear.py [hunk 7]
```diff
@@ -485,7 +486,7 @@ def forward(
         output_parallel = self.quant_method.apply(self, input_, bias)
         if self.gather_output:
             # All-gather across the partitions.
-            output = tensor_model_parallel_all_gather(output_parallel)
+            output = self.collective_func(output_parallel)
```
Comment (mgoin) [inline vllm/model_executor/layers/linear.py:489]: Why do you need to make these collective func changes to the linear layers?

File: vllm/worker/hpu_worker.py (hunk 3 above)
Comment (mgoin) [reply vllm/worker/hpu_worker.py:474]: Why do you need a new kv cache dtype "fp8_inc" for this? It seems you could decide this based on the device being on cpu

**Review by mgoin — COMMENTED** [review]
I'm very confused by what this PR is supposed to do. I'm going to list out what is unclear or seems extraneous to help convey my perspective:
* Introduces "inc" QuantizationConfig that only produces unquantized methods
* It is adding a new kv_cache_dtype "fp8_inc" that is the same as "fp8" but uses a different storage dtype (float8_e4m3fn), only to cast that dtype to the same as "fp8" (which uses "uint8")
* Changes the LoadConfig to load on CPU for "inc" quantization for an unclear reason
* Introduces new logic to the shutdown of workers, that has specific logic based on "inc" quantization being enabled

I have trouble understanding why these changes are needed to support FP8 inference on HPU. For instance, could you use an existing method for FP8 like `fp8.py` or `compressed_tensors_w8a8_fp8.py` which there are many checkpoints for?

File: docs/features/quantization/inc.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+# FP8 INC
+
+vLLM supports FP8 (8-bit floating point) weight and activation quantization using Intel® Neural Compressor (INC) on Intel® Gaudi® 2 and Intel® Gaudi® 3 AI accelerators.
+Currently, quantization is validated only in Llama models.
+
+Intel Gaudi supports quantization of various modules and functions, including, but not limited to `Linear`, `KVCache`, `Matmul` and `Softmax`. For more information, please refer to:
+[Supported Modules\\Supported Functions\\Custom Patched Modules](https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Quantization/Inference_Using_FP8.html#supported-modules).
+
+> [!NOTE]
+> Measurement files are required to run quantized models with vLLM on Gaudi accelerators. The FP8 model calibration procedure is described in the [vllm-hpu-extention](https://github.com/HabanaAI/vllm-hpu-extension/tree/main/calibration/README.md) package.
```
[the remark this answers is not in the document]
Comment (nirda7) [reply docs/features/quantization/inc.md]: Done

File: docs/features/quantization/inc.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
+(inc)=
```
[the remark this answers is not in the document]
Comment (nirda7) [reply docs/features/quantization/inc.md]: Done

File: docs/features/quantization/supported_hardware.md
[the remark this answers is not in the document]
Comment (nirda7) [reply docs/features/quantization/supported_hardware.md:1]: The X width is different than the V width - I tried my best but once it is in preview mode it should look ok.

File: vllm/model_executor/layers/quantization/inc.py (hunk 6 above)
Comment (ulivne) [reply vllm/model_executor/layers/quantization/inc.py]: INC will patch the modules with quantization logic in later stage,
this allows less dependency in INC implementation.

File: vllm/model_executor/layers/linear.py (hunk 7 above)
Comment (ulivne) [reply vllm/model_executor/layers/linear.py]: removed

File: vllm/executor/ray_distributed_executor.py (hunk 5 above)
Comment (ulivne) [reply vllm/executor/ray_distributed_executor.py]: removed

File: vllm/worker/hpu_worker.py (hunk 3 above)
Comment (ulivne) [reply vllm/worker/hpu_worker.py]: removed

File: vllm/engine/arg_utils.py [hunk 10]
```diff
@@ -973,6 +977,8 @@ def create_load_config(self) -> LoadConfig:
         return LoadConfig(
             load_format=self.load_format,
             download_dir=self.download_dir,
+            device="cpu"
```
Comment (yma11) [inline vllm/engine/arg_utils.py:1053]: If weights are firstly loaded to `cpu`, in which step it will be moved to `hpu`?

Comment (ulivne) [reply vllm/engine/arg_utils.py:1053]: they are moved by underline INC logic, after quantization to fp8

**Comment by xuechendi** [comment]
@mgoin @robertgshaw2-redhat , please help to review this PR. We want to introduce Intel Neural Compressor as one option for online quantization method in vllm.
Currently, it will be used by HPU only (HPU codes are in plugin)

File: docs/features/quantization/README.md [hunk 11]
```diff
@@ -13,6 +13,7 @@ Contents:
 - [BitBLAS](bitblas.md)
 - [GGUF](gguf.md)
 - [GPTQModel](gptqmodel.md)
+- [Inc](inc.md)
```
Comment (mgoin) [inline docs/features/quantization/README.md]: From the doc, seems like it should be INC
```suggestion
- [INC](inc.md)
```

File: vllm/config.py [hunk 12]
```diff
@@ -1446,7 +1446,7 @@ def get_and_verify_max_len(self, max_model_len: int):
 
 
 BlockSize = Literal[1, 8, 16, 32, 64, 128]
-CacheDType = Literal["auto", "fp8", "fp8_e4m3", "fp8_e5m2"]
+CacheDType = Literal["auto", "fp8", "fp8_e4m3", "fp8_e5m2", "fp8_inc"]
```
Comment (mgoin) [inline vllm/config.py:1566]: Based on the comment here, can we remove the new cache dtype now? https://github.com/vllm-project/vllm/pull/12010#discussion_r2164760581

File: vllm/model_executor/layers/quantization/inc.py
Comment (mgoin) [reply vllm/model_executor/layers/quantization/inc.py:1]: Can you please add a header to this file explaining its purpose? It is rather confusing otherwise and this is a good place to define how this quant method works

File: vllm/model_executor/model_loader/base_loader.py [hunk 13]
```diff
@@ -32,11 +35,16 @@ def load_model(self, vllm_config: VllmConfig,
...
+        load_config = vllm_config.load_config
+        load_device = device_config.device if load_config.device is None else \
+                      load_config.device
+        target_device = torch.device(load_device)
         with set_default_torch_dtype(model_config.dtype):
             with target_device:
                 model = initialize_model(vllm_config=vllm_config,
                                          model_config=model_config)
+
+            logger.info("Loading weights on %s ...", load_device)
```
Comment (mgoin) [inline vllm/model_executor/model_loader/base_loader.py]: Make this debug

File: vllm/model_executor/model_loader/weight_utils.py [hunk 14]
```diff
@@ -148,8 +148,8 @@ def get_quant_config(model_config: ModelConfig,
     quant_cls = get_quantization_config(model_config.quantization)
 
     # GGUF doesn't have config file
-    if model_config.quantization == "gguf":
-        return quant_cls.from_config({})
+    if model_config.quantization in ("gguf", "inc"):
+        return quant_cls()
```
Comment (mgoin) [inline vllm/model_executor/model_loader/weight_utils.py:156]: Is this valid for gguf ?

File: vllm/config.py (hunk 12 above)
Comment (xuechendi) [reply vllm/config.py:1566]: Since the HPU worker for v1 is moved to plugin and v0 will be deprecated soon. We want to make the map "fp8_inc to fp8_e4m3" being more visible. 

image

Alternatively, do you think we can update the mapping function above conditional like:
```
STR_DTYPE_TO_TORCH_DTYPE = {
    "half": torch.half,
    "bfloat16": torch.bfloat16,
    "float": torch.float,
    "fp8": torch.uint8,
    "fp8_e4m3": torch.uint8 if not current_platform.is_support_fp8_e4m3() else torch.float8_e4m3fn
    "fp8_e5m2": torch.uint8,
    "int8": torch.int8,
}
```

File: docs/features/quantization/README.md (hunk 11 above)
Comment (ulivne) [reply docs/features/quantization/README.md]: done

File: vllm/model_executor/model_loader/base_loader.py (hunk 13 above)
Comment (ulivne) [reply vllm/model_executor/model_loader/base_loader.py]: done

File: vllm/model_executor/model_loader/weight_utils.py (hunk 14 above)
Comment (ulivne) [reply vllm/model_executor/model_loader/weight_utils.py:156]: yes as gguf also returns always the cls straightforward
https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/quantization/gguf.py#L51

**Comment by xuechendi** [comment]
@mgoin , please help to review the PR again, we have updated the codes and resolved most of your comments.
there is one comment will need to check your opinion:
https://github.com/vllm-project/vllm/pull/12010#discussion_r2172654929

File: vllm/config.py (hunk 12 above)
Comment (mgoin) [reply vllm/config.py:1566]: It's okay, let's just keep fp8_inc then

**Review by mgoin — APPROVED** [review]
LGTM, thanks for the iterations. Please resolve the merge conflict

**Comment by xuechendi** [comment]
@robertgshaw2-redhat @simon-mo @WoosukKwon , We have got one approval from Michael
May you check the PR as well, and help to add "ready" tag if possible, thanks so much

**Comment by xuechendi** [comment]
@mgoin , thanks for the Review, all CI are passed
